### PR TITLE
expense category controller

### PR DESCRIPTION
**Pull request by diego-centella-jalafund**
# Pull Request Title
#### Resolved task/US: ([ID-8] Expense categories)
* * * *

## What's in this pull request?
The expense category endpoint, , I had changes on the program class for use the api.

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## How Has This Been Tested?
It was tested with the link of swagger doing queries for expenses categories.

## Checklist:

- [x] My code follows the style [guidelines](https://github.com/programacion-6/ExPenSeTraCkerGroUpSiX/wiki/Convetions) of this project
- [x] I have performed a self-review of my own code
- [x] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [x] I have checked my code and corrected any misspellings

### CODE REVIEW

**Comment by zrock-dev** [comment]
Could it be possible to have one small test to verify the feature works,

**Comment by zrock-dev** [comment]
The PR #9 also has created an ExpenseController

File: RestApi/Controllers/ExpenseCategoryController.cs [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+
+[ApiController]
+[Route("api/[controller]")]
+public class ExpenseCategoryController : ControllerBase
+{
+    private static List<ExpenseCategory> _defaultCategories = new List<ExpenseCategory>
+    {
+        new ExpenseCategory { Name = "Electronics" },
+        new ExpenseCategory { Name = "Transport" },
+        new ExpenseCategory { Name = "Products" }
```
Comment (zrock-dev) [inline RestApi/Controllers/ExpenseCategoryController.cs]: What about taking the default category values from the database during startup? This way, we can dynamically load those defaults while booting the app.

**Review by zrock-dev — CHANGES_REQUESTED** [review]
Good job, I only have one observation.

File: RestApi/Persistence/DataBase/ApplicationDbContext.cs [hunk 2]
```diff
@@ -4,14 +4,24 @@
 namespace RestApi.Persistence.DataBase;
 
 public class ApplicationDbContext : DbContext
+{
+    public DbSet<User> Users { get; set; }
+    public DbSet<ExpenseCategory> ExpenseCategory { get; set; }
+    public DbSet<Expense> Expenses { get; set; }
+    public DbSet<Income> Incomes { get; set; }
+
+    public ApplicationDbContext(DbContextOptions<ApplicationDbContext> options) : base(options)
```
Comment (zrock-dev) [inline RestApi/Persistence/DataBase/ApplicationDbContext.cs:13]: This method has been left empty. Wouldn't it be better to remove it?

File: RestApi/Persistence/DataBase/ApplicationDbContext.cs [hunk 3]
```diff
@@ -4,14 +4,24 @@
...
     }
+
+    protected override void OnModelCreating(ModelBuilder modelBuilder)
+    {
+        base.OnModelCreating(modelBuilder);
+
+        modelBuilder.Entity<ExpenseCategory>().HasData(
+            new ExpenseCategory { Name = "Electronics" },
+            new ExpenseCategory { Name = "Transport" },
+            new ExpenseCategory { Name = "Products" }
```
Comment (zrock-dev) [inline RestApi/Persistence/DataBase/ApplicationDbContext.cs:24]: This is a failsafe for when the DB doesn't have default categories, right?

File: RestApi/Services.cs/ExpenseCategoryService.cs [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+
+        public async Task<ExpenseCategory?> ReadAsync(Guid id)
+        {
+            return await _context.ExpenseCategory.FirstOrDefaultAsync(c => c.Id == id);
+        }
+
+        public async Task<bool> UpdateAsync(Guid id, ExpenseCategory updatedCategory)
+        {
+            var existingCategory = await _context.ExpenseCategory.FindAsync(id);
+            if (existingCategory == null) return false;
```
Comment (zrock-dev) [inline RestApi/Services.cs/ExpenseCategoryService.cs:51]: How are exceptions being handled?

**Review by zrock-dev — CHANGES_REQUESTED** [review]
Following or definition of done, please add data type validation to methods that receive data, such as: `CreateCategory`. `UpdateCategory`.

I couldn't find the endpoints of `ExpenseCategoryController`
image

**Comment by zrock-dev** [comment]
I had forgotten to build the docker image, haha. Now I can see the endpoitns.  I have some errors:
image


```
Npgsql.PostgresException (0x80004005): 42P01: relation "ExpenseCategory" does not exist

POSITION: 30
   at Npgsql.Internal.NpgsqlConnector.ReadMessageLong(Boolean async, DataRowLoadingMode dataRowLoadingMode, Boolean readingNotifications, Boolean isReadingPrependedMessage)
   at System.Runtime.CompilerServices.PoolingAsyncValueTaskMethodBuilder`1.StateMachineBox`1.System.Threading.Tasks.Sources.IValueTaskSource<TResult>.GetResult(Int16 token)
   at Npgsql.NpgsqlDataReader.NextResult(Boolean async, Boolean isConsuming, CancellationToken cancellationToken)
   at Npgsql.NpgsqlDataReader.NextResult(Boolean async, Boolean isConsuming, CancellationToken cancellationToken)
   at Npgsql.NpgsqlCommand.ExecuteReader(Boolean async, CommandBehavior behavior, CancellationToken cancellationToken)
   at Npgsql.NpgsqlCommand.ExecuteReader(Boolean async, CommandBehavior behavior, CancellationToken cancellationToken)
   at Npgsql.NpgsqlCommand.ExecuteDbDataReaderAsync(CommandBehavior behavior, CancellationToken cancellationToken)
   at Microsoft.EntityFrameworkCore.Storage.RelationalCommand.ExecuteReaderAsync(RelationalCommandParameterObject parameterObject, CancellationToken cancellationToken)
   at Microsoft.EntityFrameworkCore.Storage.RelationalCommand.ExecuteReaderAsync(RelationalCommandParameterObject parameterObject, CancellationToken cancellationToken)
   at Microsoft.EntityFrameworkCore.Query.Internal.SingleQueryingEnumerable`1.AsyncEnumerator.InitializeReaderAsync(AsyncEnumerator enumerator, CancellationToken cancellationToken)
   at Npgsql.EntityFrameworkCore.PostgreSQL.Storage.Internal.NpgsqlExecutionStrategy.ExecuteAsync[TState,TResult](TState state, Func`4 operation, Func`4 verifySucceeded, CancellationToken cancellationToken)
   at Microsoft.EntityFrameworkCore.Query.Internal.SingleQueryingEnumerable`1.AsyncEnumerator.MoveNextAsync()
   at Microsoft.EntityFrameworkCore.EntityFrameworkQueryableExtensions.ToListAsync[TSource](IQueryable`1 source, CancellationToken cancellationToken)
   at Microsoft.EntityFrameworkCore.EntityFrameworkQueryableExtensions.ToListAsync[TSource](IQueryable`1 source, CancellationToken cancellationToken)
   at RestApi.Services.ExpenseCategoryService.ReadAsync() in /src/RestApi/Services.cs/ExpenseCategoryService.cs:line 25
   at ExpenseCategoryController.GetAllCategories() in /src/RestApi/Controllers/ExpenseCategoryController.cs:line 30
   at Microsoft.AspNetCore.Mvc.Infrastructure.ActionMethodExecutor.TaskOfIActionResultExecutor.Execute(ActionContext actionContext, IActionResultTypeMapper mapper, ObjectMethodExecutor executor, Object controller, Object[] arguments)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.<InvokeActionMethodAsync>g__Awaited|12_0(ControllerActionInvoker invoker, ValueTask`1 actionResultValueTask)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.<InvokeNextActionFilterAsync>g__Awaited|10_0(ControllerActionInvoker invoker, Task lastTask, State next, Scope scope, Object state, Boolean isCompleted)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.Rethrow(ActionExecutedContextSealed context)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.Next(State& next, Scope& scope, Object& state, Boolean& isCompleted)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.<InvokeInnerFilterAsync>g__Awaited|13_0(ControllerActionInvoker invoker, Task lastTask, State next, Scope scope, Object state, Boolean isCompleted)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeFilterPipelineAsync>g__Awaited|20_0(ResourceInvoker invoker, Task lastTask, State next, Scope scope, Object state, Boolean isCompleted)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeAsync>g__Awaited|17_0(ResourceInvoker invoker, Task task, IDisposable scope)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeAsync>g__Awaited|17_0(ResourceInvoker invoker, Task task, IDisposable scope)
   at Swashbuckle.AspNetCore.SwaggerUI.SwaggerUIMiddleware.Invoke(HttpContext httpContext)
   at Swashbuckle.AspNetCore.Swagger.SwaggerMiddleware.Invoke(HttpContext httpContext, ISwaggerProvider swaggerProvider)
   at Microsoft.AspNetCore.Authorization.AuthorizationMiddleware.Invoke(HttpContext context)
   at Microsoft.AspNetCore.Authentication.AuthenticationMiddleware.Invoke(HttpContext context)
   at Microsoft.AspNetCore.Diagnostics.DeveloperExceptionPageMiddlewareImpl.Invoke(HttpContext context)
  Exception data:
    Severity: ERROR
    SqlState: 42P01
    MessageText: relation "ExpenseCategory" does not exist
    Position: 30
    File: parse_relation.c
    Line: 1191
    Routine: parserOpenTable

HEADERS
=======
Accept: */*
Connection: keep-alive
Host: localhost:5001
User-Agent: Mozilla/5.0 (X11; Linux x86_64; rv:128.0) Gecko/20100101 Firefox/128.0
Accept-Encoding: gzip, deflate, br, zstd
Accept-Language: en-US,en;q=0.5
Cookie: jenkins-timestamper-offset=14400000; structurizr.darkMode=true; screenResolution=1920x1080; JSESSIONID.cdcc472a=node015crjw3zuuskgm6l1yr5mr4h33.node0; JSESSIONID=F6CA389ABE0F30ED5A565C007A8D5367
Referer: http://localhost:5001/index.html
Sec-Fetch-Dest: empty
Sec-Fetch-Mode: cors
Sec-Fetch-Site: same-origin
DNT: 1
Sec-GPC: 1
Priority: u=0

``` 

image

**Comment by zrock-dev** [comment]
Ok, now I was able to run the project xd. I've found a bug. I was able to create a duplicated expense category
image

---
Following the bug it also affects when creating a new entry. I've sent this request and got an error
```
{
  "id": "3fa85f64-5717-4562-b3fc-2c963f66afa6",
  "amount": 0,
  "date": "2024-09-27T15:15:34.855Z",
  "description": "This is an expense",
  "category": {
    "id": "047a8117-010a-4042-8012-15eb6d76b197",
    "name": "Products"
  }
}
``` 

```
Microsoft.EntityFrameworkCore.DbUpdateException: An error occurred while saving the entity changes. See the inner exception for details.
 ---> Npgsql.PostgresException (0x80004005): 23505: duplicate key value violates unique constraint "PK_ExpenseCategory"

DETAIL: Detail redacted as it may contain sensitive data. Specify 'Include Error Detail' in the connection string to include this information.
   at Npgsql.Internal.NpgsqlConnector.ReadMessageLong(Boolean async, DataRowLoadingMode dataRowLoadingMode, Boolean readingNotifications, Boolean isReadingPrependedMessage)
   at System.Runtime.CompilerServices.PoolingAsyncValueTaskMethodBuilder`1.StateMachineBox`1.System.Threading.Tasks.Sources.IValueTaskSource<TResult>.GetResult(Int16 token)
   at Npgsql.NpgsqlDataReader.NextResult(Boolean async, Boolean isConsuming, CancellationToken cancellationToken)
   at Npgsql.NpgsqlDataReader.NextResult(Boolean async, Boolean isConsuming, CancellationToken cancellationToken)
   at Npgsql.NpgsqlCommand.ExecuteReader(Boolean async, CommandBehavior behavior, CancellationToken cancellationToken)
   at Npgsql.NpgsqlCommand.ExecuteReader(Boolean async, CommandBehavior behavior, CancellationToken cancellationToken)
   at Npgsql.NpgsqlCommand.ExecuteDbDataReaderAsync(CommandBehavior behavior, CancellationToken cancellationToken)
   at Microsoft.EntityFrameworkCore.Storage.RelationalCommand.ExecuteReaderAsync(RelationalCommandParameterObject parameterObject, CancellationToken cancellationToken)
   at Microsoft.EntityFrameworkCore.Storage.RelationalCommand.ExecuteReaderAsync(RelationalCommandParameterObject parameterObject, CancellationToken cancellationToken)
   at Microsoft.EntityFrameworkCore.Update.ReaderModificationCommandBatch.ExecuteAsync(IRelationalConnection connection, CancellationToken cancellationToken)
  Exception data:
    Severity: ERROR
    SqlState: 23505
    MessageText: duplicate key value violates unique constraint "PK_ExpenseCategory"
    Detail: Detail redacted as it may contain sensitive data. Specify 'Include Error Detail' in the connection string to include this information.
    SchemaName: public
    TableName: ExpenseCategory
    ConstraintName: PK_ExpenseCategory
    File: nbtinsert.c
    Line: 563
    Routine: _bt_check_unique
   --- End of inner exception stack trace ---
   at Microsoft.EntityFrameworkCore.Update.ReaderModificationCommandBatch.ExecuteAsync(IRelationalConnection connection, CancellationToken cancellationToken)
   at Microsoft.EntityFrameworkCore.Update.Internal.BatchExecutor.ExecuteAsync(IEnumerable`1 commandBatches, IRelationalConnection connection, CancellationToken cancellationToken)
   at Microsoft.EntityFrameworkCore.Update.Internal.BatchExecutor.ExecuteAsync(IEnumerable`1 commandBatches, IRelationalConnection connection, CancellationToken cancellationToken)
   at Microsoft.EntityFrameworkCore.Update.Internal.BatchExecutor.ExecuteAsync(IEnumerable`1 commandBatches, IRelationalConnection connection, CancellationToken cancellationToken)
   at Microsoft.EntityFrameworkCore.ChangeTracking.Internal.StateManager.SaveChangesAsync(IList`1 entriesToSave, CancellationToken cancellationToken)
   at Microsoft.EntityFrameworkCore.ChangeTracking.Internal.StateManager.SaveChangesAsync(StateManager stateManager, Boolean acceptAllChangesOnSuccess, CancellationToken cancellationToken)
   at Npgsql.EntityFrameworkCore.PostgreSQL.Storage.Internal.NpgsqlExecutionStrategy.ExecuteAsync[TState,TResult](TState state, Func`4 operation, Func`4 verifySucceeded, CancellationToken cancellationToken)
   at Microsoft.EntityFrameworkCore.DbContext.SaveChangesAsync(Boolean acceptAllChangesOnSuccess, CancellationToken cancellationToken)
   at Microsoft.EntityFrameworkCore.DbContext.SaveChangesAsync(Boolean acceptAllChangesOnSuccess, CancellationToken cancellationToken)
   at RestApi.Services.ExpenseService.CreateAsync(Expense expense) in /src/RestApi/Services.cs/ExpensesService.cs:line 19
   at RestApi.Controllers.ExpenseController.Create(Expense expense) in /src/RestApi/Controllers/ExpensesController.cs:line 29
   at Microsoft.AspNetCore.Mvc.Infrastructure.ActionMethodExecutor.TaskOfIActionResultExecutor.Execute(ActionContext actionContext, IActionResultTypeMapper mapper, ObjectMethodExecutor executor, Object controller, Object[] arguments)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.<InvokeActionMethodAsync>g__Awaited|12_0(ControllerActionInvoker invoker, ValueTask`1 actionResultValueTask)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.<InvokeNextActionFilterAsync>g__Awaited|10_0(ControllerActionInvoker invoker, Task lastTask, State next, Scope scope, Object state, Boolean isCompleted)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.Rethrow(ActionExecutedContextSealed context)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.Next(State& next, Scope& scope, Object& state, Boolean& isCompleted)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.<InvokeInnerFilterAsync>g__Awaited|13_0(ControllerActionInvoker invoker, Task lastTask, State next, Scope scope, Object state, Boolean isCompleted)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeFilterPipelineAsync>g__Awaited|20_0(ResourceInvoker invoker, Task lastTask, State next, Scope scope, Object state, Boolean isCompleted)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeAsync>g__Awaited|17_0(ResourceInvoker invoker, Task task, IDisposable scope)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeAsync>g__Awaited|17_0(ResourceInvoker invoker, Task task, IDisposable scope)
   at Swashbuckle.AspNetCore.SwaggerUI.SwaggerUIMiddleware.Invoke(HttpContext httpContext)
   at Swashbuckle.AspNetCore.Swagger.SwaggerMiddleware.Invoke(HttpContext httpContext, ISwaggerProvider swaggerProvider)
   at Microsoft.AspNetCore.Authorization.AuthorizationMiddleware.Invoke(HttpContext context)
   at Microsoft.AspNetCore.Authentication.AuthenticationMiddleware.Invoke(HttpContext context)
   at Microsoft.AspNetCore.Diagnostics.DeveloperExceptionPageMiddlewareImpl.Invoke(HttpContext context)

HEADERS
=======
Accept: */*
Connection: keep-alive
Host: localhost:5001
User-Agent: Mozilla/5.0 (X11; Linux x86_64; rv:128.0) Gecko/20100101 Firefox/128.0
Accept-Encoding: gzip, deflate, br, zstd
Accept-Language: en-US,en;q=0.5
Content-Type: application/json
Cookie: jenkins-timestamper-offset=14400000; structurizr.darkMode=true; screenResolution=1920x1080; JSESSIONID.cdcc472a=node015crjw3zuuskgm6l1yr5mr4h33.node0; JSESSIONID=F6CA389ABE0F30ED5A565C007A8D5367
Origin: http://localhost:5001
Referer: http://localhost:5001/index.html
Content-Length: 236
Sec-Fetch-Dest: empty
Sec-Fetch-Mode: cors
Sec-Fetch-Site: same-origin
DNT: 1
Sec-GPC: 1
Priority: u=0

```

File: RestApi/Persistence/DataBase/ApplicationDbContext.cs (hunk 3 above)
Comment (diego-centella-jalafund) [reply RestApi/Persistence/DataBase/ApplicationDbContext.cs:24]: They are default categories that already be on the DB

File: RestApi/Services.cs/ExpenseCategoryService.cs (hunk 4 above)
Comment (diego-centella-jalafund) [reply RestApi/Services.cs/ExpenseCategoryService.cs:51]: It was updated the exception for update categoy expense

File: RestApi/Persistence/DataBase/ApplicationDbContext.cs (hunk 2 above)
Comment (diego-centella-jalafund) [reply RestApi/Persistence/DataBase/ApplicationDbContext.cs:13]: It is the constructor for that class

File: RestApi/Persistence/DataBase/ApplicationDbContext.cs (hunk 3 above)
Comment (zrock-dev) [reply RestApi/Persistence/DataBase/ApplicationDbContext.cs:24]: nice

File: RestApi/Persistence/DataBase/ApplicationDbContext.cs (hunk 2 above)
Comment (zrock-dev) [reply RestApi/Persistence/DataBase/ApplicationDbContext.cs:13]: Ok. If the class requires it as so, then let it be

File: RestApi/Services.cs/ExpenseCategoryService.cs (hunk 4 above)
Comment (zrock-dev) [reply RestApi/Services.cs/ExpenseCategoryService.cs:51]: nice

**Review by zrock-dev — APPROVED** [review]
Good job

**Pull request merged**